MPM: cleanups

eudoxos opened the pull request:
A few things happening here:

- moving MPMs initialization away from `EngngModel` into its own class MPMSymbolicProblem (which is base for respective classes using the functionality); could be perhaps improved. I am also not sure whether I covered all of MPM, since some of the classes don't seem to be covered by tests. 
- adding `meta_enum.hpp` for named enumerations (so that one can say `quantity Pressure` instead of `quantity 3` and get a nice error message as well), plus enums used in MPM in backward-compatible manner (others can be later converted, too); this works for all readers (text, XML).
- finally `cook2_u1p0_2.xml` was updated with keywords, and (to me) it looks really nice and legible, as the structure stands out.

eudoxos commented:
(please consider upgrading to c++20 as requirement ;it includes std::format; all these printf issues would be thing of the past)

bpatzak commented:
I think it was premature, I just want to talk about the pros/cons first.  If I had to summarize the main pros/cons of the meta_enum solution: 
- meta enum is ready, perhaps effective solution 

- for me, the implementation of meta_enum is complicated and hard to understand, difficult for possible service, the authors themselves mention problems with some compilers. This is my primary concern.

- in my opinion, the original solution is simple, it has no problems with any compiler, it is basically basic c++, simple maintenance resulting potentially to extensive, but simple code logic.

eudoxos commented:
(there is also [meta_enum_lite](https://github.com/vimontgames/meta_enum_lite), which might work better for us here)

> meta_enum is complicated and hard to understand, difficult for possible service, the authors themselves mention problems with some compilers. This is my primary concern.

The complex part is parsing the enum spcification as constexpr (evaluated at compile time), the rest is fairly straightforward. Compilers required are clang 6.0, msvc >= 2018, it is all standard c++17.

The nice thing for me is that

- everything is wrapped in one function-like call, in one place. So e.g. adding string→value conversion can be done in one place (and used for all enums), currently we'd have to do it manually everywhere.
- once there is c++26 (well, future, yes, but it will happen one day), the macro can be replaced by function using reflection abilities of that language (enumerating enum names and similar).

An incremental inprovement (with all my dislikes for preprocessor, for using ENUM_ITEM and ENUM_WITH_VALUE etc) would be something like this. Note that this won't support things like namespaces (all in `oofem::`, not inside classes), `class enum` (which would be quite useful as it prohibits automatic casts to/from int).

```c
#define ENUM_TYPE ElementGeoemtryType
#define ENUM_DEF ENUM_ITEM(EGT_point) ENUM_ITEM(EGT_line_1 /* ... */ ENUM_ITEM_WITH_VALUE(FooBar, 33)
#include "enum-impl.ipp"
#undef ENUM_TYPE
#undef ENUM_DEF
```
and then `enum-impl.ipp`
```c++
// declare this somewhere
template<typename T> struct EnumTraits;
// define the enum itself
#define ENUM_ITEM(i) i,
#define ENUM_ITEM_WITH_VALUE(i,val) i=v ,
enum ENUM_TYPE{ ENUM_DEF };
#undef ENUM_ITEM_WITH_VALUE
#undef ENUM_ITEM
// conversions
template<> struct EnumTraits<ENUM_TYPE>{
   #define ENUM_ITEM_WITH_VALUE(i,val) {i,#i},
   #define ENUM_ITEM(i) {i,#i},
   static constexpr std::unordered_map<ENUM_TYPE,const char*> auto value_to_name={ ENUM_DEF };
   #undef ENUM_ITEM_WITH_VALUE
   #undef ENUM_ITEM

   #define ENUM_ITEM_WITH_VALUE(i,val) {#i,i},
   #define ENUM_ITEM(i) {#i,i},
   static constexpr std::unordered_map<const char*,ENUM_TYPE> auto name_to_value={ ENUM_DEF };
   #undef ENUM_ITEM_WITH_VALUE
   #undef ENUM_ITEM

   static std::optional<const char*> name(ENUM_NAME v){ if(auto i=value_to_name.find(v); i!=value_to_name.end()) return *i; return {} };
   static std::optional<ENUM_TYPE> value(const char* n){ if(auto i=name_to_value.find(n); i!=name_to_value.end()) return *i; return {} };
};
// compatibility with the current functions
inline const char* __ ## ENUM_TYPE ## ToString(ENUM_TYPE value){ return EnumTraits<ENUM_TYPE>::name(value).value(); }
```

bpatzak commented:
Thank you, this is understandable and simple solution. Did you considered to introduce a class able to represent the enum values but behave as a rich type that:
- Holds the enum value internally
- Converts automatically to the enum type
- Converts to/from std::string

Something like this (generated with the chatbot help):
```
// Generic wrapper for enum values
template<typename E>
class EnumValue {
    static_assert(std::is_enum_v<E>, "EnumValue can only wrap enum types");

public:
    // Constructors
    EnumValue(E value) : value_(value) {}
    EnumValue(const std::string& str) {
        if (!EnumTraits<E>::fromString(str, value_)) {
            throw std::invalid_argument("Invalid enum string: " + str);
        }
    }

    // Implicit conversion to enum
    operator E() const { return value_; }

    // Implicit conversion to string
    operator std::string() const { return EnumTraits<E>::toString(value_); }

    // Explicit getter
    E value() const { return value_; }

private:
    E value_;
};

int main() {
    try {
        EnumValue<Color> c1 = Color::Green; // from enum
        EnumValue<Color> c2 = std::string("Red"); // from string

        // Auto convert to string
        std::string s = c1;
        std::cout << "c1 as string: " << s << "\n";

        // Auto convert to enum
        Color e = c2;
        std::cout << "c2 as enum index: " << static_cast<int>(e) << "\n";

        // Invalid string throws
        EnumValue<Color> bad("Purple");
    }
    catch (const std::exception& ex) {
        std::cerr << "Error: " << ex.what() << "\n";
    }
}
```

eudoxos commented:
> Did you considered to introduce a class able to represent the enum values but behave as a rich type

[wise_enum](https://github.com/quicknir/wise_enum) discusses disadvantages of enum-like classes; the most important for us is that it cannot be used in `swich{ case ...: ...;}`. It is otherwise close to `enum class` in that all uses must be qualified (i.e. `Color::Red`, not just `Red`) and the code is not ready for that.

I added a include-file-based implementation, which is minimally intrusive, supports (optional) prefix for keywords, so that one can write `Total` instead of `VM_Total` (that is for future keyword compatibility), plus suggestion when wrong keyword is used:

```
OOFEM Error exception: quantity (enum VariableQuantity): unrecognized name 'Displcement' (did you mean 'Displacement'?)
Possible names: Displacement Velocity Temperature Pressure VolumeFraction.
```

I have no idea why the MSVC build it failing, therer are no details in the workflow log, as far as I see.


(the other commits are what I cleaned up on the way)

eudoxos commented:
(I will break this down into smaller PRs which are more manageable)
